### PR TITLE
add email distribution

### PR DESCRIPTION
**Pull request by ShirshovAlex**
add email

### CODE REVIEW

**Review by Maksim-maevsky — DISMISSED** [review]

File: pom.xml [hunk 1]
```diff
@@ -151,15 +152,31 @@
             <artifactId>jackson-datatype-jsr310</artifactId>
             <version>${jakson.datatype.version}</version>
         </dependency>
-
-
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-mail</artifactId>
+            <version>${email.version}</version>
```
Comment (SergeyKoval) [inline pom.xml]: Do you really need version here? I believe it should be extended from parent

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Properties;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class EmailServiceImpl implements EmailService {
+
+    private final EmailProperties emailProperties;
+
+    private JavaMailSender getJavaMailSender() {
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: Can't you reuse this staff?

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    public void sendSimpleMessage(FormRegisterDTO formRegisterDTO) {
+        EmailTemplate emailTemplate = new EmailTemplate();
+        SimpleMailMessage message = new SimpleMailMessage();
+        message.setFrom(emailTemplate.getFrom());
+        message.setTo(formRegisterDTO.getEmail());
+        message.setSubject(emailTemplate.getSubject());
+        message.setText(emailTemplate.getText());
+
+        getJavaMailSender().send(message);
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: Do you handling mail sending somehow?

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (ShirshovAlex) [reply pom.xml]: done

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java (hunk 2 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: done

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java (hunk 3 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: done

File: src/main/java/by/exadel/internship/mail/EmailTemplate.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@NoArgsConstructor
+public class EmailTemplate {
+    private String from = "test.exadel.test@gmail.com";
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/mail/EmailTemplate.java]: Looks like all fields in template are constants. If so make it static final and remove all annotations

**Review by Maksim-maevsky — DISMISSED** [review]

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        mailSender.setHost(emailProperties.getHost());
+        mailSender.setPort(emailProperties.getPort());
+        mailSender.setUsername(emailProperties.getUsername());
+        mailSender.setPassword(emailProperties.getPassword());
+
+        Properties props = mailSender.getJavaMailProperties();
+        props.putAll(MAIL_PROPS);
+    }
+
+    public boolean sendSimpleMessage(FormRegisterDTO formRegisterDTO) {
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: Rename to something which will identify what kind of email are you going to send. For example: sendFormSubmitionEmail

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        EmailTemplate emailTemplate = new EmailTemplate();
+        SimpleMailMessage message = new SimpleMailMessage();
+        message.setFrom(emailTemplate.getFrom());
+        message.setTo(formRegisterDTO.getEmail());
+        message.setSubject(emailTemplate.getSubject());
+        message.setText(emailTemplate.getText());
+
+        mailSender.send(message);
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: On exception it should return false and not exception

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: src/main/java/by/exadel/internship/mail/EmailTemplate.java (hunk 4 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/mail/EmailTemplate.java]: done

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java (hunk 5 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: done

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java (hunk 6 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: done

**Review by Maksim-maevsky — DISMISSED** [review]

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class EmailServiceImpl implements EmailService {
+
+    private final EmailProperties emailProperties;
+    private JavaMailSenderImpl mailSender;
+
+    private static final Map<String,String> MAIL_PROPS = Map.of("mail.transport.protocol", "smtp",
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: Why properties is not declared in `application.yaml`?

File: src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java [hunk 8]
```diff
@@ -86,7 +89,7 @@ private void uploadFile(MultipartFile multipartFile, UUID id) {
             log.info("Success to upload file, form id: {}", id);
 
         } catch (IOException e) {
-            log.error(e.getMessage());
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java]: Rollback log removal, please :-)

File: src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java [hunk 9]
```diff
@@ -69,8 +68,12 @@ private FormFullDTO saveForm(FormRegisterDTO formRegisterDTO) {
 
         formRepository.save(form);
 
-        return mapper.toFormDto(form);
+        FormFullDTO dto = mapper.toFormDto(form);
 
+        if (emailService.sendFormSubmissionEmail(formRegisterDTO)) {
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java]:
```suggestion
       dto.setConfirmEmail(emailService.sendFormSubmissionEmail(formRegisterDTO)) {
```

File: src/main/java/by/exadel/internship/dto/formDTO/FormRegisterDTO.java [hunk 10]
```diff
@@ -68,4 +68,6 @@
 
     private List<TimeForCallDTO> timeForCallList;
 
+    private boolean confirmEmail = false;
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/dto/formDTO/FormRegisterDTO.java]: JsonProperty.Access.READ_ONLY - for avoiding setting this flag from UI.

```suggestion
   @JsonProperty( access = JsonProperty.Access.READ_ONLY )
    private boolean sendEmail = false;
```

File: src/main/java/by/exadel/internship/mail/EmailTemplate.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package by.exadel.internship.mail;
+
+public class EmailTemplate {
+    public static final String FROM = "test.exadel.test@gmail.com";
+    public static final String SUBJECT = "Test Exadel";
+    public static final String TEXT = "Your form has been successfully submitted.";
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/mail/EmailTemplate.java]: As I remember you would like to add name here?

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java (hunk 7 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: done

File: src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java (hunk 8 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java]: done

File: src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java (hunk 9 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java]: done

File: src/main/java/by/exadel/internship/dto/formDTO/FormRegisterDTO.java (hunk 10 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/dto/formDTO/FormRegisterDTO.java]: done

File: src/main/java/by/exadel/internship/mail/EmailTemplate.java (hunk 11 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/mail/EmailTemplate.java]: done

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class EmailServiceImpl implements EmailService {
+
+    private final JavaMailSenderImpl mailSender;
+
+    public boolean sendFormSubmissionEmail(FormRegisterDTO formRegisterDTO) {
+        SimpleMailMessage message = new SimpleMailMessage();
+        String text = EmailTemplate.TEXT.replace("${name}", formRegisterDTO.getFirstName());
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: Move "${name}" to constants

File: src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java [hunk 13]
```diff
@@ -32,13 +33,13 @@
 @Slf4j
 public class FormServiceImpl implements FormService {
 
+    private static final String SIMPLE_CLASS_NAME = FormService.class.getSimpleName();
```
Comment (SergeyKoval) [inline src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java:36]: Why do you need it. Looks strange

**Review by SergeyKoval — CHANGES_REQUESTED** [review]

File: src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java (hunk 13 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/FormServiceImpl.java:36]: that's not mine. Nikita did it for log4j

File: src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java (hunk 12 above)
Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/service/impl/EmailServiceImpl.java]: done

File: src/main/java/by/exadel/internship/config/EmailConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.mail.javamail.JavaMailSenderImpl;
+
+import java.util.Map;
+import java.util.Properties;
+
+@Configuration
+@RequiredArgsConstructor
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/config/EmailConfiguration.java]:
```suggestion
```

Comment (ShirshovAlex) [reply src/main/java/by/exadel/internship/config/EmailConfiguration.java]: done

**Review by SergeyKoval — APPROVED** [review]

**Review by kzaretskaya — APPROVED** [review]

**Pull request merged**